sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetime.scala

```diff
@@ -0,0 +1,375 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import java.sql.Date
+import java.text.SimpleDateFormat
+
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.expressions.codegen.{GeneratedExpressionCode, CodeGenContext}
+import org.apache.spark.sql.catalyst.util.DateTimeUtils
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.UTF8String
+
+abstract class DateFormatExpression extends Expression { self: Product =>
+  
+  protected val date: Expression
+  
+  protected val format: Expression
+
+  override def foldable: Boolean = date.foldable && format.foldable
+
+  override def nullable: Boolean = true
+
+  override def children: Seq[Expression] = Seq(date, format)
+
+  override def eval(input: InternalRow): Any = {
+    val valueLeft = date.eval(input)
+    if (valueLeft == null) {
+      null
+    } else {
+      val valueRight = format.eval(input)
+      if (valueRight == null) {
+        null
+      } else {
+        val sdf = new SimpleDateFormat(valueRight.asInstanceOf[UTF8String].toString)
+        date.dataType match {
+          case TimestampType =>
+            UTF8String.fromString(sdf.format(new Date(valueLeft.asInstanceOf[Long] / 10000)))
+          case DateType =>
+            UTF8String.fromString(sdf.format(DateTimeUtils.toJavaDate(valueLeft.asInstanceOf[Int])))
+          case StringType =>
+            UTF8String.fromString(
+              sdf.format(DateTimeUtils.stringToTime(valueLeft.asInstanceOf[UTF8String].toString)))
+        }
+      }
+    }
+  }
+
+  /**
+   * Called by date format expressions to generate a code block that returns the result
+   *
+   * As an example, the following parse the result to int
+   * {{{
+   *   defineCodeGen(ctx, ev, c => s"Integer.parseInt($c.toString())")
+   * }}}
+   *
+   * @param f function that accepts a variable name and returns Java code to parse an
+   *          [[UTF8String]] to the expected output type
+   */
+
+  protected def defineCodeGen(
+      ctx: CodeGenContext,
+      ev: GeneratedExpressionCode,
+      f: String => String): String = {
+
+    val sdf = "java.text.SimpleDateFormat"
+    val utf8 = "org.apache.spark.unsafe.types.UTF8String"
+    val dtUtils = "org.apache.spark.sql.catalyst.util.DateTimeUtils"
+
+    val eval1 = date.gen(ctx)
+    val eval2 = format.gen(ctx)
+
+    val varName = ctx.freshName("resultVar")
+
+    s"""
+      ${eval1.code}
+      boolean ${ev.isNull} = ${eval1.isNull};
+      ${ctx.javaType(dataType)} ${ev.primitive} = ${ctx.defaultValue(dataType)};
+      if (!${ev.isNull}) {
+        ${eval2.code}
+        if (!${eval2.isNull}) {
+          $sdf sdf = new $sdf(${eval2.primitive}.toString());
+          Object o = ${eval1.primitive};
+          $utf8 $varName;
+          if (o instanceof ${ctx.boxedType(TimestampType)}) {
```

Comment:
Yes this would be possible, if the type of the left parameter type never changes. For dataframe this is true, for Spark SQL I'm not sure. Can I just use pattern Scala matching to decide which transformation is correct, like

```
eval2 match {
  case StringType => ...
  case ....
}
```